### PR TITLE
[package-extractor] Add file paths to extractor-metadata.json

### PR DESCRIPTION
**Pull request by D4N14L**
## Summary

Add a "files" field to the extractor-metadata.json file generated as part of extraction. Also perform some cleanup and deduplication of code.

## How it was tested

Improved existing tests and validated new output.

## Impacted documentation

N/A

### CODE REVIEW

File: common/changes/@rushstack/package-extractor/user-danade-AddFileTracking_2024-09-14-08-27.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@rushstack/package-extractor",
+      "comment": "Add a \"files\" field to the extractor-metadata.json file",
```
Comment (iclanton) [inline common/changes/@rushstack/package-extractor/user-danade-AddFileTracking_2024-09-14-08-27.json:5]:
```suggestion
      "comment": "Add a `files` field to the `extractor-metadata.json` file",
```

File: libraries/package-extractor/src/scripts/create-links.ts [hunk 2]
```diff
@@ -121,23 +120,28 @@ function main(): boolean {
...
-  process.exitCode = 1;
-  if (main()) {
-    process.exitCode = 0;
-  }
-} catch (error) {
-  console.log('ERROR: ' + error);
-}
+process.exitCode = 1;
+const terminal: Terminal = new Terminal(new ConsoleTerminalProvider());
+main(terminal)
```
Comment (iclanton) [inline libraries/package-extractor/src/scripts/create-links.ts:141]:
```suggestion
runAsync(terminal)
```

File: libraries/package-extractor/src/scripts/create-links.ts [hunk 3]
```diff
@@ -3,112 +3,111 @@
...
+      terminal.writeVerboseLine(`Realizing file at path "${filePath}"`);
+      if (process.platform === 'win32') {
+        await FileSystem.createHardLinkAsync({ newLinkPath: filePath, linkTargetPath: realFilePath });
       } else {
-        FileSystem.createSymbolicLinkFile({ newLinkPath, linkTargetPath });
+        await FileSystem.copyFileAsync({ sourcePath: realFilePath, destinationPath: filePath });
       }
     }
-  }
+  });
```
Comment (iclanton) [inline libraries/package-extractor/src/scripts/create-links.ts:90]: Limit concurrency?

File: libraries/package-extractor/src/AssetHandler.ts [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+      createArchiveOnly = false
+    } = options;
+    this._terminal = terminal;
+    this._sourceRootFolder = sourceRootFolder;
+    this._targetRootFolder = targetRootFolder;
+    this._symlinkAnalyzer = symlinkAnalyzer;
+    if (createArchiveFilePath) {
+      if (path.extname(createArchiveFilePath) !== '.zip') {
+        throw new Error('Only archives with the .zip file extension are currently supported.');
+      }
```
Comment (iclanton) [inline libraries/package-extractor/src/AssetHandler.ts:70]: It'd be great to add support for tarballs too.

File: libraries/package-extractor/src/AssetHandler.ts [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        return;
+      }
+      throw new Error(`The asset at path "${targetFilePath}" has already been included`);
+    }
+
+    if (!this._createArchiveOnly) {
+      // Ignore when the source file is the same as the target file, as it's a no-op
+      if (sourceFilePath && sourceFilePath !== targetFilePath) {
+        // Use the fs.copyFile API instead of FileSystem.copyFileAsync() since copyFileAsync performs
+        // a needless stat() call to determine if it's a file or folder, and we already know it's a file.
```
Comment (iclanton) [inline libraries/package-extractor/src/AssetHandler.ts:111]: Can probably update this to try to copy as a file, and then fall back to copying as a folder if that fails.

**Review by iclanton — APPROVED** [review]

File: libraries/package-extractor/src/AssetHandler.ts (hunk 4 above)
Comment (D4N14L) [reply libraries/package-extractor/src/AssetHandler.ts:70]: Agree, though I think that would be a future PR/out of scope for this one.

File: libraries/package-extractor/src/AssetHandler.ts (hunk 5 above)
Comment (D4N14L) [reply libraries/package-extractor/src/AssetHandler.ts:111]: Idk, I actually think if the target is a directory that should fail. The alternate source option (`sourceFileContent`) can only ever be a file. Seems appropriate for the `EISDIR` error to be thrown

**Pull request merged**